---
title: "DNA Methylation Analysis"
author: "Kristin"
date: "11 7 2022"
output: pdf_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning = FALSE, message = FALSE)
```
```{r libraries}
library(dplyr)
library(BiocManager)
library(TCGAbiolinks)
library(SummarizedExperiment)
library(limma)
library(minfi)
library(readr)
library(ExperimentHub)
library(DMRcate)
library(RColorBrewer)
library(stringr)
library(ggplot2)
library(EnhancedVolcano)
library(biomaRt)
library(coMET)
library(annotatr)

```

## Download the methylation (beta-values) data from TCGA using biolinks


```{r download, eval = FALSE}
mdna_query <- GDCquery(project = "TCGA-LIHC",
                       data.category = "DNA Methylation",
                       data.type = "Methylation Beta Value",
                       platform = "Illumina Human Methylation 450")


GDCdownload(mdna_query, method = "api", files.per.chunk = 50,
            directory = "biolinksData/")

data <- GDCprepare(query = mdna_query, save = TRUE,  save.filename = "dnaM_client.rda",  directory = "clientData/", summarizedExperiment = TRUE)

```

```{r loadObject, include = FALSE}
load("data/dnaM.rda")
```


## Add clinical information
```{r clinical}
met <- as.data.frame(SummarizedExperiment::assay(data))
clinical = colDataPrepare(data@colData$samples)
clinical = clinical[clinical$patient %in%  clinical[clinical$sample_type=="Solid Tissue Normal", "patient"],]
```

## Data Preprocessing

Load CpG annotation information for HM450 array on hg38 
```{r hg38_anno}
hg38_anno = read_tsv("data/HM450.hg38.manifest.tsv.gz")
```

Remove probes with missing values
```{r na_removal}
probe.na <- rowSums(is.na(met))
table(probe.na == 0)
probe <- probe.na[probe.na == 0]
met <- met[row.names(met) %in% names(probe), ]
```

Remove probes mapping to X and Y to remove sex bias
```{r sex_chr}
keep <- !(row.names(met) %in% hg38_anno$probeID[hg38_anno$CpG_chrm %in% c("chrX","chrY")])
met <- met[keep, ]
table(keep) 
```

Remove probes overlapping SNPs with MAF > 0.01
```{r snp_removal}
no.snp.probe <- hg38_anno$probeID[!(hg38_anno$MASK_snp5_GMAF1p)]
met <- met[row.names(met) %in% c(no.snp.probe), ]
table(hg38_anno$MASK_snp5_GMAF1p)

```

```{r clean, echo = FALSE}
rm(probe.na)
rm(probe)
rm(keep)
rm(no.snp.probe)
```

Count tissue distribution and check ordering of samples
```{r order}
table(clinical$sample_type)
met = met[,row.names(clinical)]
```

Transform beta values into m values 
```{r mval}
mval <- t(apply(met, 1, function(x) log2(x/(1-x))))
```

## Exploratory Data Analysis

Beta value distribution
```{r}
densityPlot(as.matrix(met), sampGroups = clinical$sample_type)
```
A perfect beta distribution would only contain two peaks around 0 and 1. The peaks between 0 and 1 indicate that there is heterogeneity within some tumor samples. The distribution of beta-values within the normal samples looks good, indicating tissue homogeneity. 

MDS plot
```{r}
pal <- brewer.pal(8,"Dark2")
limma::plotMDS(met, top=1000, gene.selection="common", 
        col=pal[factor(clinical$sample_type)], labels = NULL, pch = 19)
```
The normal tissue samples (orange) cluster well together regarding both components. Unfortunately, there also is a lot of heterogeneity between tumor samples causing a widespread distribution in regard to both components. 

## Differential methylated CpG loci (DMLs)

Define the design and fit the model
```{r}
sample_type = factor(clinical$sample_type)

design <- model.matrix(~sample_type)
fit <- lmFit(mval, design)
fit <- eBayes(fit)
```

Get the summary statistics
```{r}
infinite = topTable(fit, number = 288556)
head(infinite)
```
Volcano plot of CpG loci
```{r}
EnhancedVolcano(infinite,
                lab = rownames(infinite),
                x = 'logFC',
                y = 'adj.P.Val')
```
Filter the loci by adjusted p-value (<0.005) and mean beta value difference between groups (>0.2)
```{r}
diff = data.frame(MeanBeta_diff = rowMeans(met[,sample_type=="Solid Tissue Normal"]) - rowMeans(met[,sample_type=="Primary Tumor"]))
merged_stats = merge(infinite, diff, by = 0)
row.names(merged_stats) = merged_stats[,1]
merged_stats = merged_stats[,-1]
merged_stats = merged_stats %>% filter(adj.P.Val < 0.005 & MeanBeta_diff > 0.2) %>% arrange(adj.P.Val)
head(merged_stats)
cat("Number of differentially methylated loci: ", nrow(merged_stats))
```

Plot the 4 most differentially methylated loci
```{r}
par(mfrow=c(2,2))
sapply(rownames(merged_stats)[1:4], function(cpg){
  plotCpg(met, cpg=cpg, pheno=sample_type, ylab = "Beta values")
})
```
## Differential methylated regions (DMRs)

Analyse differential methylation of CpGs. Attention: annotates CpG loci to hg19!
```{r}
myannotation <- cpg.annotate("array", mval, what = "M", arraytype = "450K",
                             analysis.type="differential", design=design, coef=ncol(design), fdr = 0.01)
```

Replace coordinates in myannotation\@ranges object to hg38 coordinates
```{r}
true_coord = hg38_anno[,c(1,2,3,5)]
anno_df = data.frame(myannotation@ranges)
anno_df$probeID = names(myannotation@ranges)
merged_df = merge(true_coord, anno_df)
merged_df = merged_df[,c(1,2,3,9,10,11,12,13)]
merged_df$start = merged_df$CpG_beg + 1
merged_df$end = merged_df$CpG_beg + 1
Cpg_IDs = merged_df$probeID
merged_df = merged_df[,-c(1,3)]
colnames(merged_df) = c("seqnames","strand","stat","diff","ind.fdr","is.sig", "start", "end")
true_GRanges = makeGRangesFromDataFrame(merged_df, keep.extra.columns = T)  
myannotation@ranges = true_GRanges
names(myannotation@ranges) = Cpg_IDs
```

Run dmrcate to combine differential methlyated CpG loci to regions
```{r}
dmrcoutput <- dmrcate(myannotation, lambda=1000, C=2, betacutoff = 0.2)

results.ranges <- extractRanges(dmrcoutput, genome = "hg38")
results.ranges
```
Helper function to visualize DMRs as genomic region track
```{r}
visualize_dmr = function(dmrIndex, gen, Granges, file){
  
  cat(dmrIndex, "\n")

  # coordinates are stored under results.ranges[dmrIndex]
  chrom <- as.character(seqnames(results.ranges[dmrIndex]))
  start <- as.numeric(start(results.ranges[dmrIndex]))
  end <- as.numeric(end(results.ranges[dmrIndex]))
  
  # add 25% extra space to plot
  minbase <- start - (0.25*(end-start))
  maxbase <- end + (0.25*(end-start))
  
  
  # defining CpG islands track
  islandTrack = cpgIslands_UCSC(gen, chrom, start, end, title="CpG Islands UCSC")
  
  #Setting up the ideogram, genome, and RefSeq tracks 
  
  iTrack <- IdeogramTrack(genome = gen, chromosome = chrom, name=paste0(chrom))
  gTrack <- GenomeAxisTrack(col="black", cex=1, name="", fontcolor="black")
  biomTrack <- BiomartGeneRegionTrack(genome = gen,
                                      chromosome = chrom, start = minbase, end = maxbase,
                                      name = "ENSEMBL Genes", collapseTranscripts = "longest")
  
  #Ensure that the methylation data is ordered by chromosome and base position.
  bvalOrd <- met[names(myannotation@ranges),]
  
  #Create the data tracks:
  #create genomic ranges object from methylation data
  cpgData <- GRanges(myannotation@ranges, beta = bvalOrd)
  
  #methylation data track
  methTrack <- DataTrack(range=cpgData, 
                         groups=sample_type,
                         genome = gen,
                         chromosome=chrom,
                         ylim=c(-0.05,1.05),
                         col=pal,
                         type=c("a","p"), 
                         name="DNA Meth.\n(beta value)",
                         background.panel="white", 
                         legend=TRUE, 
                         cex.title=0.8,
                         cex.axis=0.8, 
                         cex.legend=0.8)
  
  #DMR position data track
  dmrTrack <- AnnotationTrack(start=start, end=end, genome=gen, name="DMR", 
                              chromosome=chrom,fill="darkred")
  
  
  # Set up the tracklist and indicate the relative sizes of the different tracks. 
  # Finally, draw the plot using the plotTracks function
  tracks <- list(iTrack, gTrack, methTrack, dmrTrack, islandTrack,biomTrack)
  sizes <- c(2,2,7,2,2,2) # set up the relative sizes of the tracks
  # to save figure and scaling graphic device
  #pdf(file)
  plotTracks(tracks, from=minbase, to=maxbase, showTitle=TRUE, add53=TRUE, 
             add35=TRUE, grid=TRUE, lty.grid=3, sizes = sizes, length(tracks), transcriptAnnotation = "symbol")
  #dev.off()
}
```

Visualize some DMRs
```{r}
for(i in c(2,3,5,8)){
  visualize_dmr(i,"hg38", results.ranges, paste0("plots/DMR_new", i, ".pdf"))
}
```

Get hyper- and hypomethlated regions 
```{r}
hyper_regions = results.ranges[results.ranges$meandiff<0,]
hypo_regions = results.ranges[results.ranges$meandiff>0,]

```


Get available annotations for hg38
```{r}
annos = builtin_annotations()[unlist(lapply(builtin_annotations(), grepl, pattern = "hg38"))]
annos
annos_genes = build_annotations("hg38", c("hg38_genes_promoters", "hg38_genes_exons", "hg38_genes_introns", "hg38_genes_5UTRs", "hg38_genes_3UTRs", "hg38_genes_intergenic"))
annos_cpgs = build_annotations("hg38", c("hg38_cpg_islands", "hg38_cpg_shores", "hg38_cpg_shelves", "hg38_cpg_inter"))

```

Annotate hypermethylated regions
```{r}
annotated_dmrs_genes = annotate_regions(regions = hyper_regions, annotations = annos_genes)
plot_annotation(annotated_regions = annotated_dmrs_genes)

annotated_dmrs_cpgs = annotate_regions(regions = hyper_regions, annotations = annos_cpgs)
plot_annotation(annotated_regions = annotated_dmrs_cpgs) 
```

Annotate hypomethylated regions
```{r}
annotated_dmrs_genes = annotate_regions(regions = hypo_regions, annotations = annos_genes)
plot_annotation(annotated_regions = annotated_dmrs_genes)

annotated_dmrs_cpgs = annotate_regions(regions = hypo_regions, annotations = annos_cpgs)
plot_annotation(annotated_regions = annotated_dmrs_cpgs)

```


