---
title: "DNAm"
author: "Kristin"
date: "28 6 2022"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message = FALSE, warning = FALSE)
```

## Differential DNA methylation analysis


```{r libraries}
library(dplyr)
library(BiocManager)
library(TCGAbiolinks)
library(SummarizedExperiment)
library(limma)
library(readr)
library(wateRmelon)
library(ExperimentHub)
library(DMRcate)
library(RColorBrewer)
library(minfi)
library(DMRcate)
```


```{r pressure, echo=FALSE}
setwd("C:/Users/Krissi/Desktop/TCGA_LIHC/LIHC-project/")
```

## Download methylation data (beta values) from TCGA
```{r}
#mdna_query <- GDCquery(project = "TCGA-LIHC",
#                       data.category = "DNA Methylation",
#                       data.type = "Methylation Beta Value",
#                       platform = "Illumina Human Methylation 450")


#GDCdownload(mdna_query, method = "api", files.per.chunk = 50,
#            directory = "biolinksData/")

#data <- GDCprepare(query = mdna_query, save = TRUE,  save.filename = "dnaM_client.rda",  directory = "clientData/", summarizedExperiment = TRUE)
```

```{r, echo = FALSE}
##load data
load("data/dnaM.rda")
```

## Download basic clinical data from TCGA

```{r}
##extract cases for which both normal and tumor tissue is available
met <- as.data.frame(SummarizedExperiment::assay(data))
clinical = colDataPrepare(data@colData$samples)
clinical = clinical[clinical$patient %in%  clinical[clinical$sample_type=="Solid Tissue Normal", "patient"],]
```

## Download more clinical data from TCGA
```{r}
barcodes = data@colData$samples
barcodes = unlist(lapply(barcodes, substr, start = 1, stop = 12))

clinical_query <- GDCquery(project = "TCGA-LIHC",
                           data.category = "Clinical", 
                           barcode = barcodes)

GDCdownload(clinical_query, method = "api", files.per.chunk = 100,
            directory = "ClinicalData/")

info <- GDCprepare_clinic(query = clinical_query, clinical.info = "patient", directory = "ClinicalData/")
info.stage_event <- GDCprepare_clinic(query = clinical_query, clinical.info = "stage_event", directory = "ClinicalData/")

info = merge(info, info.stage_event, by = 'bcr_patient_barcode')
```

## Preprocess methylation data
```{r}
# get  450k annotation data for hg38
hg38_anno = read_tsv("data/HM450.hg38.manifest.tsv.gz")

```


```{r}
## remove probes with NA
probe.na <- rowSums(is.na(met))

cat("rows without missing values: \n")
table(probe.na == 0)

#choose those without NA values in rows
probe <- probe.na[probe.na == 0]
met <- met[row.names(met) %in% names(probe), ]

##remove probes that match to chromosome  X and Y 
keep <- !(row.names(met) %in% hg38_anno$probeID[hg38_anno$CpG_chrm %in% c("chrX","chrY")])
cat("CpG's not on X or Y chrom: \n")
table(keep)

met <- met[keep, ]

## remove SNPs overlapped probe with MAF > 0.01
cat("CpG's overlapping common SNPs: \n")
table(hg38_anno$MASK_snp5_GMAF1p)

# probes without snp with MAF > 0.01
no.snp.probe <- hg38_anno$probeID[!(hg38_anno$MASK_snp5_GMAF1p)]

# filter met
met <- met[row.names(met) %in% c(no.snp.probe), ]

#remove no-further needed dataset
rm(probe.na)
rm(probe)
rm(keep)
rm(no.snp.probe)


#intra-sample normalisation procedure, correcting the bias of type-2 probe values
#BMIQ(met, design.v)
##get design.v?
```
## EDA
```{r}
densityPlot(as.matrix(met), sampGroups = clinical$sample_type)

pal <- brewer.pal(8,"Dark2")
plotMDS(met, top=1000, gene.selection="common", 
        col=pal[factor(clinical$sample_type)], labels = NULL, pch = 19)
```

## Differential methylation analyses
```{r}
cat("Number of samples per tissue: \n")
table(clinical$sample_type)

##check ordering
met = met[,row.names(clinical)]

design <- model.matrix(~ clinical$sample_type)
fit <- lmFit(met, design)
fit <- eBayes(fit)
top_10 = topTable(fit)

# plot the top 4 most significantly differentially methylated CpGs 
par(mfrow=c(2,2))
sapply(rownames(top_10)[1:4], function(cpg){
  plotCpg(met, cpg=cpg, pheno=clinical$sample_type, ylab = "Beta values")
})


myannotation <- cpg.annotate("array", as.matrix(met), what = "Beta", arraytype = "450K",
                             analysis.type="differential", design=design, coef=2, fdr = 0.01)

dmrcoutput <- dmrcate(myannotation, lambda=2000, C=2)

results.ranges <- extractRanges(dmrcoutput, genome = "hg38")
DMR.plot(ranges=results.ranges, dmr=1, CpGs=as.matrix(met), what="Beta",
         arraytype = "450", phen.col=clinical$sample_type, genome="hg38")

```

